doc/content/xcp-networkd/host-network-device-ordering.md

```diff
@@ -0,0 +1,181 @@
+host should connect to.
+
+At first glance, this could be resolved by specifying a mapping between an
+individual network device and a pool-wide network. However, this approach
+would be burdensome for administrators when managing many hosts. It would be
+more efficient if the Toolstack could determine this automatically.
+
+To achieve this, the Toolstack components on two hosts need to independently
+work out consistent identifications for the host network devices and connect
+the network devices with the same identification to the same pool-wide network.
+The identifications on a host can be considered as an order, with each network
+device assigned a unique position in the order as its identification. Network
+devices with the same position will connect to the same network.
+
+
+The assumption
+--------------
+
+Why can the Toolstack components on two hosts independently work out an expected
+order without any communication? This is possible only under the assumption that
+the hosts have identical hardware, firmware, software, and the way
+network devices are plugged into them. For example, an administrator will always
+plug the network devices into the same PCI slot position on multiple hosts if
+they want these network devices to connect to the same network.
+
+The ordering is considered consistent if the positions of such network devices
+(plugged into the same PCI slot position) in the generated orders are the same.
+
+Particularly, when this assumption holds, a consistent initial order can be
+worked out on multiple hosts independently with the help of `biosdevname`. The
+"all_ethN" policy of the `biosdevname` utility can generate a device order based
+on whether the device is embedded or not, PCI cards in ascending slot order, and
+ports in ascending PCI bus/device/function order breadth-first. Since the hosts
+are identical, the orders generated by the `biosdevname` are consistent across
+the hosts.
+
+
+The initial order
+-----------------
+
+As mentioned above, the `biosdevname` can be used to generate consistent orders
+for the network devices on multiple hosts. However, this only works if the
+assumption of identical hardware, firmware, software, and network device
+placement holds. When this assumption doesn't hold, the order can be generated
+from a user's configuration. However, administrators would prefer to avoid this
+as much as possible when managing many hosts.
+
+
+Keep the order as stable as possible
+------------------------------------
+
+Once an initial order is created on an individual host, it should be kept as
+stable as possible across host boot-ups and at runtime. For example, unless
+there are hardware changes, the position of a network device in the initial
+order should remain the same regardless of how many times the host is rebooted.
+
+To achieve this, the initial order should be saved persistently on the host's
+local storage so it can be referenced in subsequent orderings. When performing
```

Comment:
Are we going to use file based storage or use the xapi database for this? Given that this is needed early in system start, xapi could be not available or at least in a failure case - such that working with files could be easier.

Reply:
Yes, file based storage is to be used. The networkd doesn't use xapi.

Reply:
I would suggest to use a single file for this to avoid inconsistencies. Could be an easy to parse line-oriented format or JSON.